### PR TITLE
[SPARK-19382][ML]:Test sparse vectors in LinearSVCSuite

### PR DESCRIPTION
**Pull request by wangmiao1981**
## What changes were proposed in this pull request?

Add unit tests for testing SparseVector.

We can't add mixed DenseVector and SparseVector test case, as discussed in JIRA 19382.

 def merge(other: MultivariateOnlineSummarizer): this.type = {
if (this.totalWeightSum != 0.0 && other.totalWeightSum != 0.0) {
require(n == other.n, s"Dimensions mismatch when merging with another summarizer. " +
s"Expecting $n but got $
{other.n}

.")

## How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by wangmiao1981** [comment]
Jenkins, retest this please.

**Comment by wangmiao1981** [comment]
retest this please

**Comment by wangmiao1981** [comment]
Jenkins, retest this please.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala [hunk 1]
```diff
@@ -220,12 +246,13 @@ object LinearSVCSuite {
     "aggregationDepth" -> 3
   )
 
-    // Generate noisy input of the form Y = signum(x.dot(weights) + intercept + noise)
+  // Generate noisy input of the form Y = signum(x.dot(weights) + intercept + noise)
   def generateSVMInput(
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala]: This API is strange, where the caller expects numFeatures = weights.size, but really numFeatures = 10 * weights.size if isDense=false.  Please update it to construct a random dense or sparse vector first (both of length weights.size) and then compute y to make the API more consistent.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala [hunk 2]
```diff
@@ -234,7 +261,12 @@ object LinearSVCSuite {
       val yD = new BDV(xi).dot(weightsMat) + intercept + 0.01 * rnd.nextGaussian()
       if (yD > 0) 1.0 else 0.0
     }
-    y.zip(x).map(p => LabeledPoint(p._1, Vectors.dense(p._2)))
+    val index = (0 to weights.length - 1).toArray
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala]: Move inside if-then to branch where it is used.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala [hunk 3]
```diff
@@ -203,6 +227,8 @@ class LinearSVCSuite extends SparkFunSuite with MLlibTestSparkContext with Defau
     val svm = new LinearSVC()
     testEstimatorAndModelReadWrite(svm, smallBinaryDataset, LinearSVCSuite.allParamSettings,
       checkModelData)
+    testEstimatorAndModelReadWrite(svm, smallSparseBinaryDataset, LinearSVCSuite.allParamSettings,
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala]: No need for this.  Once the model has been fit, its training data is irrelevant.

**Review by jkbradley — COMMENTED** [review]
@wangmiao1981  Thanks for the patch!  I don't think that we need to replicate all tests using dense and sparse vectors.  The main concern is making sure that sparse and dense vectors produce the same behavior.

Is this multivariate online summarizer issue really a bug?  Or is it from passing in sparse vectors which are 10x longer than the dense vectors?

I'd recommend:
* Change the dense and sparse datasets so that they match:
  * create the sparse datasets
  * convert the sparse ones to dense using Vector.toDense
* Create a test suite which trains on the 2 datasets and confirms that the models produced are exactly the same.
* Revert the other test changes.  Unless we can think of reasons why the aspect being tested will interact with vector sparsity, then we don't need to add tests.

**Comment by wangmiao1981** [comment]
@jkbradley I simplified the tests and modified the data generation API by using toSparse method, which eliminates the index variable.
"Is this multivariate online summarizer issue really a bug? Or is it from passing in sparse vectors which are 10x longer than the dense vectors?"
It is not a bug based on my understanding. It is because the sparse size is larger than the dense vector size, which means we can't mix using them together.

**Comment by wangmiao1981** [comment]
@jkbradley I simplified the test cases and modified the data generation API using to toSparse method, which eliminates the index variable. 
"Is this multivariate online summarizer issue really a bug? Or is it from passing in sparse vectors which are 10x longer than the dense vectors?"
I think it is not a bug, which is due to the different sizes between dense and sparse vectors. It means that we can't mix using the two together.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala [hunk 4]
```diff
@@ -68,13 +77,21 @@ class LinearSVCSuite extends SparkFunSuite with MLlibTestSparkContext with Defau
     val model = svm.fit(smallBinaryDataset)
     assert(model.transform(smallValidationDataset)
       .where("prediction=label").count() > nPoints * 0.8)
+    val sparseModel = svm.fit(smallSparseBinaryDataset)
+    assert(sparseModel.transform(smallSparseValidationDataset)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala]: No need to do transform for this model; calling checkModels should suffice.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala [hunk 5]
```diff
@@ -68,13 +77,21 @@ class LinearSVCSuite extends SparkFunSuite with MLlibTestSparkContext with Defau
...
+    checkModels(model, sparseModel)
   }
 
   test("Linear SVC binary classification with regularization") {
     val svm = new LinearSVC()
     val model = svm.setRegParam(0.1).fit(smallBinaryDataset)
     assert(model.transform(smallValidationDataset)
       .where("prediction=label").count() > nPoints * 0.8)
+    val sparseModel = svm.fit(smallSparseBinaryDataset)
+    assert(sparseModel.transform(smallSparseValidationDataset)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala]: same here

File: mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala [hunk 6]
```diff
@@ -51,6 +54,12 @@ class LinearSVCSuite extends SparkFunSuite with MLlibTestSparkContext with Defau
     smallBinaryDataset = generateSVMInput(A, Array[Double](B, C), nPoints, 42).toDF()
     smallValidationDataset = generateSVMInput(A, Array[Double](B, C), nPoints, 17).toDF()
     binaryDataset = generateSVMInput(1.0, Array[Double](1.0, 2.0, 3.0, 4.0), 10000, 42).toDF()
+
+    // Dataset for testing SparseVector
+    smallSparseBinaryDataset = generateSVMInput(A, Array[Double](B, C), nPoints, 42, false).toDF()
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LinearSVCSuite.scala]: Why call generateSVMInput again?  It seems brittle.  I'd prefer to use a UDF to convert the vectors to sparse vectors here.

**Review by jkbradley — COMMENTED** [review]
Thanks for the simplification!  A few comments.

**Comment by jkbradley** [comment]
Thanks for the updates.  This LGTM pending the conflict resolution.  Sorry for the delay!

**Comment by wangmiao1981** [comment]
Resolved. Thanks!

**Comment by jkbradley** [comment]
LGTM
Thanks @wangmiao1981 !
Merging with master

**Pull request closed without merging**